Blog: Added filter by tags

NLTN opened the pull request:
* Created TagField Component
	File Location: src/lib/components/utils/acm-tagfield.svelte
	Usage: Allows users to make selections and filter content.

* Created new file utils.ts to store common functions
	File Location: src/lib/common/utils.ts

* Added Filter by Tags feature for Blog page.
	Made some changes to these files:
	- In blog/index.svelte: Added TagField component.
	- In blog/_query.ts: Modified fetchNewsletters.
	- In blog/index.json.ts: Modified get function.

This PR is intended to resolve #277

NLTN committed:
Added Filter by Tag feature

NLTN committed:
Added TagField component

NLTN committed:
Updated TagField on:change event  & Fixed minor bugs

NLTN committed:
Updated TagField + Added discernTags() + Modified fetchNewsletters()

TagField can work with Javascript disabled.

NLTN committed:
Optimized discernTags()

EthanThatOneKid commented on src/routes/blog/index.svelte:
```diff
@@ -1,9 +1,25 @@
 <script lang="ts" context="module">
   import type { LoadInput, LoadOutput } from '@sveltejs/kit/types/internal';
+  import TagField from '$lib/components/utils/acm-tagfield.svelte';
```
Nit: Import the `TagField` component along with the rest of the component import statements in the 2nd `<script>`.

NLTN replied:
Good catch. Thanks.

EthanThatOneKid commented on src/routes/blog/index.xml.ts:
```diff
@@ -53,7 +53,8 @@ function makeRssFeed(posts: Newsletter[]): string {
 
 export async function get(): Promise<RequestHandlerOutput> {
-  return new Response(JSON.stringify(makeRssFeed(await fetchNewsletters())), {
+  const posts: Newsletter[] = (await fetchNewsletters()).posts;
```
Nit: Fancy [_deconstructor syntax_](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Destructuring_assignment) simplification.
```suggestion
  const { posts } = await fetchNewsletters();
```

EthanThatOneKid commented on src/lib/components/utils/acm-tagfield.svelte:
```diff
@@ -0,0 +1,129 @@
+      }
+    } else {
+      return '';
+    }
```
## Suggestions for `createTagURL()`

### 1) Nit: To avoid bugs, let's make the condition in the if-statement more strict.

```ts
if (urlSearchParamKey === '') {
```

### 2) To keep from making too many indents, let's short circuit.

```ts
if (urlSearchParamKey === '') {
  return '';
}
```

### 3) Let's determine the `nextTags` to be selected before constructing the string.

```ts
const nextTags = selectedTags.includes(tag)
      ? selectedTags.filter((t) => t !== tag)
      : selectedTags.concat([tag]);
```

### 4) Let's take advantage of the `URLSearchParams` API.

```ts
const params = [[urlSearchParamKey, nextTags.join(',')]];
return '?' + new URLSearchParams(params).toString();
```

### 5) Now, let's put it all together.

```suggestion
    if (urlSearchParamKey === '') {
      return '';
    }
    const nextTags = selectedTags.includes(tag)
      ? selectedTags.filter((t) => t !== tag)
      : selectedTags.concat([tag]);
    const params = [[urlSearchParamKey, nextTags.join(',')]];
    return '?' + new URLSearchParams(params).toString();
```

EthanThatOneKid commented on src/routes/blog/_query.ts:
```diff
@@ -103,6 +113,14 @@ export async function fetchNewsletters(): Promise<Newsletter[]> {
+  let newsletters = formatNewsletters(await response.json());
+  const tags = discernTags(newsletters);
+
+  if (options && options.labels.length > 0) {
```
Nit: Strictly assert that the `options` object is defined.
```suggestion
  if (options !== undefined && options.labels.length > 0) {
```

jaasonw replied:
Alternatively, just `if (options?.labels.length > 0)`

jaasonw replied:
More info: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Optional_chaining

> The ?. operator is like the . chaining operator, except that instead of causing an error if a reference is [nullish](https://developer.mozilla.org/en-US/docs/Glossary/Nullish) ([null](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/null) or [undefined](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/undefined)), the expression short-circuits with a return value of undefined. When used with function calls, it returns undefined if the given function does not exist.

NLTN replied:
Thank you @EthanThatOneKid for pointing out. 
I like your solution @jaasonw. Tested and it worked as expected. Thanks

EthanThatOneKid commented on src/routes/blog/_query.ts:
```diff
@@ -103,6 +113,14 @@ export async function fetchNewsletters(): Promise<Newsletter[]> {
 
-  const newsletters = formatNewsletters(await response.json());
-  return newsletters;
+  let newsletters = formatNewsletters(await response.json());
```
Since we are expecting to return a property `posts`, let's name this variable `posts` for convenience.
```suggestion
  const posts = formatNewsletters(await response.json());
```

NLTN replied:
I used `let` because it will be reassigned at line 120.

jaasonw replied:
oop- yea i just noticed LOL I didnt read that far down yet

EthanThatOneKid replied:
Good catch! Sounds good.

EthanThatOneKid commented on src/routes/blog/_query.ts:
```diff
@@ -103,6 +113,14 @@ export async function fetchNewsletters(): Promise<Newsletter[]> {
+    );
+  }
+
+  return { tags: tags, posts: newsletters };
```
To be fancy, let's use this object construction shorthand syntax:
```suggestion
  return { tags, posts };
```

EthanThatOneKid commented on src/routes/blog/_query.ts:
```diff
@@ -103,6 +113,14 @@ export async function fetchNewsletters(): Promise<Newsletter[]> {
+  const tags = discernTags(newsletters);
+
+  if (options && options.labels.length > 0) {
+    newsletters = newsletters.filter((post) =>
```
Do not forget to swap out the variable names in this line.
```suggestion
    posts = posts.filter((post) =>
```

EthanThatOneKid commented on src/lib/common/utils.ts:
```diff
@@ -0,0 +1,11 @@
+  labels: string[];
+}
+
+export function discernTags(posts: WithLabels[]): string[] {
```
Let's create a couple test cases ([_example_](https://github.com/EthanThatOneKid/acmcsuf.com/blob/dcd14855e9e14b402b2c72d41f16a0b26d5f7bef/src/lib/common/utils.test.ts#L4)) in `/src/lib/common/utils.test.ts` for `discernTags()` to make sure we have the behavior we want.

EthanThatOneKid commented on src/routes/blog/index.svelte:
```diff
@@ -1,9 +1,25 @@
-    const response = await fetch(`/blog.json`);
-    return { props: { posts: await response.json() } };
+  let tags = [];
+  let selectedTags: string[] = [];
```
These two variables should not be reachable from the `module` script. The `module` script is ran on the server and sometimes on the browser. Therefore, we want to remove the variable declarations here and move them down to the 2nd `<script>`.
```suggestion
```

NLTN replied:
Yup, I will fix it. Now I feel like I begin to understand Svelte.

EthanThatOneKid commented on src/routes/blog/index.svelte:
```diff
@@ -1,9 +1,25 @@
+  let tags = [];
+  let selectedTags: string[] = [];
+
+  export async function load(event: LoadInput): Promise<LoadOutput> {
```
Nit: Since we only use `event` to access `event.url`, let's deconstruct `url` out right away.
```suggestion
  export async function load({ url }: LoadInput): Promise<LoadOutput> {
```

NLTN replied:
I fixed it.

EthanThatOneKid commented on src/routes/blog/index.svelte:
```diff
@@ -1,9 +1,25 @@
+    const blogOutput = await response.json();
+    tags = blogOutput.tags;
+
+    return { props: { posts: blogOutput.posts } };
```
We must pass information to the client by returning it as a returned property of the `load` function.
```suggestion
    const target = new URL('/blog.json', url);
    const rawLabels = url.searchParams.get('l');
    
    const selectedTags = [];
    if (rawLabels !== null && rawLabels.length > 0) {
      target.searchParams.set('l', rawLabels);
      selectedTags.push(...rawLabels.split(','));
    }

    const response = await fetch(target.toString());
    const { posts, tags } = await response.json();

    return { props: { posts, tags, selectedTags } };
```

NLTN replied:
I fixed it.

EthanThatOneKid commented on src/routes/blog/index.svelte:
```diff
@@ -12,10 +28,23 @@
   import Spacing from '$lib/components/sections/spacing.svelte';
 
   export let posts: Newsletter[] = [];
```
Expect there to be more properties passed down.
```suggestion
  export let posts: Newsletter[] = [];
  export let tags: string[] = [];
  export let selectedTags: string[] = [];
```

NLTN replied:
Yup, I will fix it. Now I feel like I begin to understand Svelte.

EthanThatOneKid commented on src/routes/blog/index.svelte:
```diff
@@ -12,10 +28,23 @@
+  async function filterPosts(event: CustomEvent) {
+    const tags = event.detail;
+
+    window.history.replaceState({}, '', `${window.location.pathname}?l=${tags.join(',')}`);
```
Nit: `window` APIs are accessible without the `window` object.
```suggestion
    history.replaceState({}, '', location.pathname + '?l=' + tags.join(','));
```

NLTN replied:
I fixed it.

EthanThatOneKid commented on src/routes/blog/index.svelte:
```diff
@@ -12,10 +28,23 @@
+
+    const response = await fetch(target.toString());
+    const blogOutput = await response.json();
+    posts = blogOutput.posts;
```
What if this request fails?

NLTN replied:
I fixed it.

EthanThatOneKid reviewed (changes requested):
This PR represents a very complicated and well thought out feature. I left several suggestions regarding Svelte/TypeScript techniques. 

Do not hesitate to ask for clarification for any of the comments I left. Please reply to each comment with something like "Done." so that I can resolve the individual comment if it is ready to be resolved.

NLTN committed:
Update src/lib/components/utils/acm-tagfield.svelte

Great suggestion. Much cleaner code & avoid bugs.

Co-authored-by: Ethan Davidson <31261035+EthanThatOneKid@users.noreply.github.com>

jaasonw commented on src/routes/blog/index.json.ts:
```diff
@@ -1,8 +1,14 @@
+export async function get(event: RequestEvent): Promise<RequestHandlerOutput> {
+  const fetchOptions: NewsletterFetchOptions = { labels: [] };
+
+  if (event.url.searchParams.has('l') === true && event.url.searchParams.get('l').length > 0) {
```
I'm of the opinion that checking === true is redundant, just do `if (event.url.searchParams.has('l') && event.url.searchParams.get('l').length > 0)`

conveys the same information whilst being more concise

NLTN replied:
> I'm of the opinion that checking === true is redundant, just do `if (event.url.searchParams.has('l') && event.url.searchParams.get('l').length > 0)`
> 
> conveys the same information whilst being more concise

You’re right. I fixed it couple weeks ago.

NLTN committed:
Update src/routes/blog/index.xml.ts

OMG. Deconstructor syntax is so helpful.

Co-authored-by: Ethan Davidson <31261035+EthanThatOneKid@users.noreply.github.com>

NLTN committed:
Fix bugs and improved code quality.

NLTN committed:
Created a unit test for discernTags() function

NLTN committed:
Added Svelte Slot component to TagField component

EthanThatOneKid commented on src/lib/common/utils.test.ts:
```diff
@@ -0,0 +1,16 @@
+      { labels: ['kilo', 'victor'] },
+      { labels: ['delta', 'alpha', 'bravo'] },
+    ])
+  ).toEqual(['alpha', 'bravo', 'delta', 'george washington', 'kilo', 'victor']);
```
1. The name of the unit test should describe the behavior we are checking for.
2. We should break up the unit tests to make them more readable.
3. We should use simpler sample data to test with (which makes it easier to understand at a glance).

Below is a unit test that tests the behavior expected on line 7 (testing the case where multiple labels are discerned in alphabetical order). The unit test below can be used as an example to make separate unit tests for the expectations on lines 5 (testing the case where the argument has no labels) and 6 (testing the case where 1 label is discerned).

```suggestion
test('discerns unique labels from list in alphabetical order', () => {
  const input = [
    { labels: [] },
    { labels: ['abc def'] },
    { labels: ['d', 'c', 'c', 'c'] },
    { labels: ['e', 'd'] },
    { labels: ['c', 'a', 'b'] },
  ];
  const expected = ['a', 'abc def', 'b', 'c', 'd', 'e'];
  const output = utils.discernTags(input);
  expect(output).toEqual(expected);
```

**Note**: Make sure you have 2 more unit test definitions representing the expectations on lines 5 and 6.

NLTN replied:
The unit test has been updated.

EthanThatOneKid commented on src/lib/components/utils/acm-tagfield.svelte:
```diff
@@ -0,0 +1,128 @@
+    <slot name="title" />
+  </div>
+  <a href="?" class="tag-clear-button" class:hidden={!hasSelectedTags} on:click={deselectAll}>
+    <slot name="resetButton" />
```
For slot names, I do not believe we currently have decided on a naming convention. I personally think `name="reset-button"` with a `-` (hyphen; kebab-case) looks looks nicer in HTML markup. 
```suggestion
    <slot name="reset-button" />
```

EthanThatOneKid replied:
There are currently [no occurrences](https://cs.github.com/EthanThatOneKid/acmcsuf.com?q=%3Cslot) of slots with multi-part names in this repository (this would be the first). So to be clear, we will use kebab-case for the slot names when we encounter a multi-part slot name.

NLTN replied:
Absolutely agree. It looks nicer.

NLTN committed:
Code refactoring. Change the word "Tag" with "Label".

NLTN committed:
Optimized filterPosts()

NLTN committed:
Changed slot name.

EthanThatOneKid commented on src/routes/blog/index.json.ts:
```diff
@@ -1,8 +1,14 @@
+
+  if (event.url.searchParams.has('l') === true && event.url.searchParams.get('l').length > 0) {
+    fetchOptions.labels = event.url.searchParams.get('l').split(',');
+  }
```
Nit: Simplification of conditional statement.
```suggestion
  const fetchOptions: NewsletterFetchOptions = { labels: [] };
  const rawLabels = event.url.searchParams.get('l');

  if (rawLabels?.length > 0) {
    fetchOptions.labels = rawLabels.split(',');
  }
```

EthanThatOneKid commented on src/routes/blog/index.svelte:
```diff
@@ -1,21 +1,48 @@
+    const rawLabels = url.searchParams.get('l');
+    const selectedLabels = [];
+
+    if (rawLabels !== null && rawLabels.length > 0) {
```
Nit: Simplification of conditional statement. 
```suggestion
    if (rawLabels?.length > 0) {
```

EthanThatOneKid commented on src/routes/blog/index.xml.ts:
```diff
@@ -53,7 +53,8 @@ function makeRssFeed(posts: Newsletter[]): string {
 export async function get(): Promise<RequestHandlerOutput> {
-  return new Response(JSON.stringify(makeRssFeed(await fetchNewsletters())), {
+  const { posts } = await fetchNewsletters();
+  return new Response(JSON.stringify(makeRssFeed(posts)), {
```
Is `JSON.stringify` necessary on line 57?

NLTN replied:
Removed unnecessary code on line 57.

EthanThatOneKid commented on src/lib/components/utils/acm-labelfield.svelte:
```diff
@@ -0,0 +1,128 @@
+        display: none;
+      }
+    }
+  }
```
Nit: SCSS style definition spacing to ease the eyes (you may [hopefully] just directly commit this suggested change).
```suggestion
  .label-box {
    display: flex;
    align-items: center;
    justify-content: center;
    margin-bottom: 0.2em;

    .title {
      font-size: 1em;
      font-weight: bold;
      margin-right: 1em;
      cursor: default;
    }

    .hidden {
      display: none;
    }
    
    .reset-button {
      margin-right: 1em;
      font-size: small;
      text-decoration: underline;
      cursor: pointer;
    }
    
    .label-list {
      display: flex;
      flex-wrap: wrap;
      justify-content: center;
      margin-bottom: 0.2em;
      
      .label {
        margin-bottom: 0.2em;
        margin-right: var(--size-sm);
        padding: 0.25em 0.8em 0.25em 0.8em;
        background-color: #eeeeee;
        color: #212121;
        border-radius: var(--size-sm);
        border: 2px solid #e0e0e0;
        cursor: pointer;
        transition: 0.25s ease-in-out;
        text-decoration: none;
        
        &:hover {
          border-color: #b0bec5;
        }
        
        &.selected {
          transition: 1s ease-in-out;
          background-color: #81d4fa;
          border-color: #4fc3f7;
          
          &:before {
            content: '✓ ';
          }
          
          &:hover {
            border-color: #2196f3;
          }
        }
      }
    }
    
    @media screen and (max-width: 900px) {
      .title {
        display: none;
      }
      
      .reset-button {
        display: none;
      }
    }
  }
```

EthanThatOneKid reviewed (commented):
The code in this PR overall looks amazing! I left a couple new nitpicks related to @jaasonw's <https://github.com/EthanThatOneKid/acmcsuf.com/pull/466#discussion_r870944856> and one simply making a suggestion to space out your SCSS blocks. I cannot wait to see the search bar in production :) Getting this feature to work with JavaScript disabled was no easy task!

NLTN committed:
Update src/lib/components/utils/acm-labelfield.svelte

Co-authored-by: Ethan Davidson <31261035+EthanThatOneKid@users.noreply.github.com>

NLTN committed:
Update src/routes/blog/index.svelte

Co-authored-by: Ethan Davidson <31261035+EthanThatOneKid@users.noreply.github.com>

NLTN committed:
Remove unnecessary code

NLTN committed:
Update src/routes/blog/index.json.ts

Co-authored-by: Ethan Davidson <31261035+EthanThatOneKid@users.noreply.github.com>

NLTN committed:
Merge branch 'main' into fix/277

NLTN committed:
Fixed lint errors

NLTN committed:
Hide LabelField if there is no posts

jaasonw reviewed (changes requested):
It does not appear to be workin

![image](https://user-images.githubusercontent.com/8981287/174703517-c73c1590-acc7-47d9-b2fa-aaf698170c52.png)

NLTN commented on src/routes/blog/index.json.ts:
```diff
@@ -1,15 +1,28 @@
+      status: 200,
+      headers: { 'Content-Type': 'application/json' },
+    }
+  );
```
@jaasonw Could you please check the Environment Variables in Vercel? I think we need to have these variables: VITE_GH_ACCESS_TOKEN, VITE_GH_DISCUSSION_CATEGORY_ID, and DEBUG. I don’t have a permission to check it.

jaasonw replied:
Pinging @EthanThatOneKid for this

jaasonw commented on src/routes/blog/index.svelte:
```diff
@@ -1,24 +1,51 @@
+    }
+
+    const response = await fetch(target.toString());
+    const { posts, labels } = await response.json();
```
@NLTN Looks like console.logging the result of these variables returns 
```{ posts: undefined, labels: undefined }```
And that response only contains a list of blog posts with no labels. Did you maybe forget to commit a file?

NLTN replied:
@jaasonw I double checked, and I believe that I committed all the necessary files. BTW, everything should work smoothy no matter what a list of blog posts has labels or not.

EthanThatOneKid committed:
Merge branch 'main' into fix/277

EthanThatOneKid committed:
Fixed a mistake made when merging conflicts

Somehow, I managed to empty out the contents of `/src/lib/common/utils.ts`. I put them back, merged.

EthanThatOneKid committed:
Added support for the "no posts" UI state.

EthanThatOneKid committed:
Removed unused import to pass the tests!

EthanThatOneKid committed:
Merge branch 'main' into fix/277

EthanThatOneKid committed:
Refactored `fetchNewsletters` to handle DEBUG mode logic

NLTN committed:
Made blog page to work with/without Javascript.

EthanThatOneKid committed:
Merge branch 'main' into fix/277

EthanThatOneKid committed:
Revert "Revert "Refactored caching logic for individual blog posts (#515)" (#524)"

This reverts commit 7db1f3b4ec9d4c9406b7dbc7ee92a773541597eb.

EthanThatOneKid committed:
Fixed typo in blog page title

[4:48 PM]jasonw: i forgot to put post but
[4:48 PM]jasonw: change blog back to readme

jaasonw reviewed (approved):
👍
